### ISSUE TITLE
[UC] Administrative assistant working for department head

### ISSUE THREAD

**Issue by alanhkarp** [w3c/lws-ucs#10]
**As a department manager (Manager) I want my administrative assistant (Assistant) to exercise some, but not all, of my permissions, such as updating my calendar and screening my email, so that I don't have to waste my time on these matters.**

## Preconditions:

*What conditions must be in place or assumed before this use case can begin?*

Manager: Has a set of permissions granted because of specific attributes of having the job
Assistant: Has a set of permissions granted because of specific attributes of having the job
Services used by Manager and Assistant must be aware of the permission mechanism, or there must be a proxy service in place.

## Trigger:

*What (user or system) event or action initiates this use case?*

Manager gets a new Assistant

## Actors:

*Describe the primary actor, and any other relevant actors involved in this use case*

Manager: Has critical responsibilities, such as budget, performance evaluations, meeting scheduling, responding to telephone and email, among others

Assistant: Has non-critical responsibilities, such as answering telephone calls, ordering office supplies, scheduling department meetings

Others who may need to use select Manager or Assistant privileges on an ad hoc basis.

## Distinction:

*What unique challenges or distinguishing factors (like technical issues, user experience needs, workflow integration, etc.) are associated with this use case?*

Manager spends a lot of time scheduling appointments and dealing with unimportant emails.  Common permission infrastructure, such a Role-Based, Attribute-Based, or Policy-Based access control, doesn't provide sufficient flexibility for Manager to offload some of this work to Assistant and Assistant to offload some of that work to others.

## Scenario:

*Describe an ideal or happy-case scenario where this use case would play out as intended.*

When Manager gets a new Assistant, Manager delegates to Assistant permission to update Manager's calendar and to read and delete email Assistant deems superfluous but not to send emails as Manager.  Audit records show when Assistant uses a permission delegated by Manager.

## Alternative case(s):

*What alternative flows or variations should the system handle for this use case?*

Manager may want to delegate certain of Manager's permissions to direct reports other than Assistant.

Manager may want to delegate to Assistant additional permissions on a temporary basis at times when IT support is not available to implement policy changes.

Assistant may need to delegate some of Manager permissions on a temporary basis to someone else.

Manager may want to delegate certain permissions to someone in another organization, such as a contractor or business partner, even when there is not a formal federation.

## Error scenario:

*What unexpected issues or errors might arise, and how should the system handle them?*

Assistant may intentionally or unintentionally abuse granted permissions, such as deleting important emails.  Manager retains right to revoke permissions.

Assistant may change jobs, which Manager deals with by revoking granted permissions.

Manager may need to revoke Manager permissions that Assistant delegated to someone else without stopping Assistant from using those permissions.

## Acceptance Criteria:

*What conditions or criteria must be met for this use case to be considered successfully handled?* 

Assistant is granted the fewest privileges needed to do the job Manager needs done.

Assistant may take on duties from someone other than Manager without interfering with permissions from Manager.

Some Assistant tasks may require native Assistant permissions along with one or more permissions delegated by Manager.

*What limitations are acceptable?*

It is often hard to define the exact permissions Assistant will need, so some over provisioning is acceptable.

Manager may be unavailable to revoke permissions, so timeouts can be used as a backup.

## References:

*List any relevant resources or examples that could inform this use case, possibly from other domains or solutions.*

[OAuth 2](https://oauth.net/2/) authorization tokens
[GNAP](https://oauth.net/gnap/) certificates
[UCAN](https://github.com/ucan-wg/spec) certificates

**Comment by bumblefudge** [comment]
On the one hand, it might be tempting to delegate/punt this to a distinct ZCap/OCap/GNAP/AuthZ layer, but I think it's a good reminder, even if these use cases don't resonate as first-order for this group, that all these use-cases require transitive delegation and non-ACL delegation to "as-yet unknown parties" to be both actionable and auditable-- the manager needs to be able to audit the work that the assistant or the contractor does, the audit log needs to be aware of delegations and delegees, etc.

**Comment by alanhkarp** [comment]
Absolutely right, @bumblefudge.  The ability to audit is one of the advantages of a capability-based approach.  When delegation is inconvenient, people resort to credential sharing, which makes responsibility tracking much harder.

There is an important distinction that can affect the technology you choose for a given deployment.  A certificate-based system typically supports offline delegation, so auditing can only be done on use.  A system with opaque tokens relies on token exchange, which gives you the opportunity to audit at the time of delegation.  

Your comment about "as-yet unknown parties" points out another important aspect of such a system.  The identity of the contractor may not be meaningful to Manager, so Manager will hold Assistant responsible for use of the permission.  It's up to Assistant to hold the contractor responsible for actions taken with the delegated permission.

**Comment by uvdsl** [comment]
I would like to second this use case. It is a good example for delegation of rights between agents. In this case, the two agents are natural persons.

To further add to the story here, I would like to contribute a related use case where the delegating agent (delegator) is an organisation, a legal person (not a natural person). For example, a CEO of a company does not have all rights just by default, i.e. the CEO _is not_ the company. Technically, the CEO is only a delegate of the company, and acts on behalf of the company. 

We addressed this kind of use case in the MANDAT Project, and our colleagues published what they came up with:
Sebastian Schmid, Daniel Schraudner, Andreas Harth: [_The Rights Delegation Proxy: An Approach for Delegations in the Solid Dataspace_](https://ceur-ws.org/Vol-3705/paper02.pdf). SDS@ESWC 2024

Maybe this can also serve as input (or an additional reference) for this use case.

At this point, I am unsure if the two use cases are similar enough to be grouped here, or distinct enough to be separate use cases. There _is_ a subtle difference in the delegation between a legal person (which can never act on its own) and natural person compared to delegation between two natural persons. @hzbarcea, could you chime in with your opinion on this?

**Comment by alanhkarp** [comment]
@uvdsl I believe the cases are similar enough. 

There are many cases where delegation does not involve people, such as when enforcing least privilege.  For example, the Chrome browser protects you by running its rendering engine with a subset of your permissions.  

The part missing from the referenced example is identifying the party (natural person) who approved the policy and should be held responsible for any errors.

**Comment by hzbarcea** [comment]
> I would like to second this use case. It is a good example for delegation of rights between agents. In this case, the two agents are natural persons.
> 
> To further add to the story here, I would like to contribute a related use case where the delegating agent (delegator) is an organisation, a legal person (not a natural person). For example, a CEO of a company does not have all rights just by default, i.e. the CEO _is not_ the company. Technically, the CEO is only a delegate of the company, and acts on behalf of the company.
> 
> We addressed this kind of use case in the MANDAT Project, and our colleagues published what they came up with: Sebastian Schmid, Daniel Schraudner, Andreas Harth: [_The Rights Delegation Proxy: An Approach for Delegations in the Solid Dataspace_](https://ceur-ws.org/Vol-3705/paper02.pdf). SDS@ESWC 2024
> 
> Maybe this can also serve as input (or an additional reference) for this use case.
> 
> At this point, I am unsure if the two use cases are similar enough to be grouped here, or distinct enough to be separate use cases. There _is_ a subtle difference in the delegation between a legal person (which can never act on its own) and natural person compared to delegation between two natural persons. [@hzbarcea](https://github.com/hzbarcea), could you chime in with your opinion on this?

I agree that this is similar, yet distinct, use case. The difference between in delegations is actually not subtle at all. Using your definition of a 'legal person', one that can never act on its own, the delegation graph may become quite sophisticated, and an audit trail will be necessary, imho.

I'll take the following concrete scenario: I want to install solar panel on my roof, I identified a company/contractor to do the job, they want access to my building's blueprints. I can share access, but I don't know exactly to whom, and it should not be a natural person (may not be just one, I may not know who the person is, person my switch jobs, etc, etc). I don't know how the contractor org is structured either, so I can share with the org, but the org (an admin) may refine/forward the share to a specific department, say 'planning', so that members of the 'planning' sub-group have access to my blueprints. One some natural person accesses the data, it must be auditable why.

Btw, I have some comments related to the paper you linked. I am not sure if/how to share them.

**Comment by alanhkarp** [comment]
Yes, @hzbarcea, delegation graphs can become quite sophisticated, and many of those actions are taken by or on behalf of a legal person, not a physical one.  The ability to audit is critical in order to identify the party responsible for a given action.  

> I'll take the following concrete scenario: I want to install solar panel on my roof, I identified a company/contractor to do the job, they want access to my building's blueprints. I can share access, but I don't know exactly to whom, and it should not be a natural person (may not be just one, I may not know who the person is, person my switch jobs, etc, etc). I don't know how the contractor org is structured either, so I can share with the org, but the org (an admin) may refine/forward the share to a specific department, say 'planning', so that members of the 'planning' sub-group have access to my blueprints. One some natural person accesses the data, it must be auditable why.

That's pretty much how it should work, but let's walk through your example step by step.  

You have permission to read and write your building's blueprints.  The company has a designated agent (physical or legal) for handling new work.  You delegate read permission to your blueprints to that agent.  That agent, knowing the organization, delegates to the correct agent within the company.  That agent may delegate to another agent inside or outside the company.  Eventually someone reads your blueprints, and you ultimately get a bid for the work.  At no time do you need to know anything about the company's internal organization or if any part of the work was outsourced to a third party.  

If anything goes wrong, your recourse is with the company's agent that you first contacted.  It's up to the company to hold the next agent in the delegation chain responsible.  Note that there need not be a globally accessible audit trail.  All that's needed is for each delegator be able to identify its delegate.

Did that answer your question?

You can contact me at my [public email address]( https://github.com/alanhkarp).

**Comment by uvdsl** [comment]
> I agree that this is similar, yet distinct, use case. The difference between in delegations is actually not subtle at all. Using your definition of a 'legal person', one that can never act on its own, the delegation graph may become quite sophisticated, and an audit trail will be necessary, imho.

Okay, thank you @hzbarcea, do you want me to quickly open another issue for this use case specifically?

> >  I'll take the following concrete scenario: I want to install solar panel on my roof, I identified a company/contractor to do the job, they want access to my building's blueprints. I can share access, but I don't know exactly to whom, and it should not be a natural person (may not be just one, I may not know who the person is, person my switch jobs, etc, etc). I don't know how the contractor org is structured either, so I can share with the org, but the org (an admin) may refine/forward the share to a specific department, say 'planning', so that members of the 'planning' sub-group have access to my blueprints. One some natural person accesses the data, it must be auditable why.

> That's pretty much how it should work [...] .

Well, that is pretty much how it works in our setup / project. The company itself as a legal entity has a WebID 
(e.g. http://example.org/acme#com) and access to the blueprints is set for that company. Within the company's systems, there are policies (etc) that define which employee is allowed to handle a new case e.g. to assign another clerk or another craftsman to your contract (e.g. http://example.org/alice#me).

So, in our case it is different from what @alanhkarp wrote:
>  The company has a designated agent (physical or legal) for handling new work. You delegate read permission to your blueprints to that agent. 

This is not the case for us - there is no externally known designated agent to handle new work that access rights are set for.
In our case, the company itself as a legal person is given access, and the legal person can be held accountable by the customer.
First contact for the customer is always the contractor company, the legal person. Not an agent/employee working for the company.
So, I would disagree with the statement:
> If anything goes wrong, your recourse is with the company's agent that you first contacted.

The recourse is with the company.
If the solar panels come loose and damage the roof, the customer would sue the company and not directly the employee who installed the panels on behalf of the company. (in a simple case, there may be other instances)

I think, however, that a company may employ such a "new case agent" to screen new work and then to assign the correct clerk or craftsman.

> At no time do you need to know anything about the company's internal organization or if any part of the work was outsourced to a third party

This is correct, and I'd like to emphasize that this is also the case in our setup: The customer only needs to know the WebID of the company. Not even a single employee/agent of the company needs to be known.

> Btw, I have some comments related to the paper you linked. I am not sure if/how to share them.

The authors are best reached via email, feel free to put me in CC.
If you would like to discuss the paper directly with me, I look forward to receiving your email.

**Comment by alanhkarp** [comment]
@uvdsl, your explanation of the initial contact and the way responsibility is assigned is more precise than mine.  Both are properly the legal entity of the company, not any agent acting on the company's behalf.  There will likely be an agent on first contact, but as far as the customer is concerned that agent is the company.  It's up to the company to distinguish one of its agents from another, not the customer.
